File: gfapy/field/oriented_identifier_gfa2.py
import gfapy
import re

def unsafe_decode(string):
  return gfapy.OrientedLine(string[:-1], string[-1])

def decode(string):
  obj = unsafe_decode(string)
  validate_decoded(obj)
  return obj

def validate_encoded(string):
  if not re.match("^[!-~]+[+-]$", string):
    raise gfapy.FormatError(
      "{} is not a valid oriented GFA2 identifier\n".format(repr(string))+
      "(it contains spaces or non-printable characters, or a wrong orientation)")

def validate_decoded(obj):
  if isinstance(obj, gfapy.OrientedLine):
    if not re.match("^[!-~]+$", obj.name):
      raise gfapy.FormatError(
          "{} is not a valid oriented GFA2 identifier\n".format(repr(obj.name)))
    if obj.orient != "+" and obj.orient != "-":
      raise gfapy.FormatError(
          "{} is not a valid orientation\n".format(repr(obj.orient)))
  else:
    raise gfapy.TypeError(
      "the class {} is incompatible with the datatype\n"
      .format(obj.__class__.__name__)+
      "(accepted classes: gfapy.OrientedLine)")

def unsafe_encode(obj):
  if isinstance(obj, str):
    return obj
  if isinstance(obj, gfapy.OrientedLine):
    return str(obj)
  else:
    raise gfapy.TypeError(
      "the class {} is incompatible with the datatype\n"
      .format(obj.__class__.__name__)+
      "(accepted classes: str, gfapy.OrientedLine)")

def encode(obj):
  string = unsafe_encode(obj)
  validate_encoded(string)
  return string

File: lib/rgfa/field/oriented_identifier_gfa2.rb
module RGFA::Field::OrientedIdentifierGFA2

  def unsafe_decode(string)
    string.to_oriented_line
  end

  def decode(string)
    object = unsafe_decode(string)
    validate_decoded(object)
    return object
  end

  def validate_encoded(string)
    if string !~ /^[!-~]+[+-]$/
      raise RGFA::FormatError,
        "#{string.inspect} is not a valid GFA2 oriented identifier\n"+
        "(it contains spaces or non-printable characters,"+
        "or does not end with an orientation symbol)"
    end
  end

  def validate_decoded(object)
    case object
    when RGFA::OrientedLine
      if object.name !~ /^[!-~]+$/
        raise RGFA::ValueError,
          "#{object.inspect} is not a valid GFA2 oriented identifier\n"+
          "(segment name contains spaces or non-printable characters)"
      elsif ![:+,:-].include?(object.orient)
        raise RGFA::ValueError,
          "#{object.inspect} is not a valid GFA2 oriented identifier\n"+
          "(orientation is invalid)"
      end
    else
      raise RGFA::TypeError,
        "the class #{object.class} is incompatible with the datatype\n"+
        "(accepted classes: RGFA::OrientedLine)"
    end
  end

  def unsafe_encode(object)
    case object
    when String, RGFA::OrientedLine
      return object.to_s
    else
      raise RGFA::TypeError,
        "the class #{object.class} is incompatible with the datatype\n"+
        "(accepted classes: String, RGFA::OrientedLine)"
    end
  end

  def encode(object)
    object.kind_of?(String) ?
      validate_encoded(object) : validate_decoded(object)
    return object.to_s
  end

  module_function :decode
  module_function :unsafe_decode
  module_function :validate_encoded
  module_function :validate_decoded
  module_function :unsafe_encode
  module_function :encode

end
